### ISSUE TITLE
Build Official ARM64 Images

### ISSUE THREAD

**Issue by lgadban** [solo-io/gloo#3486]
**Is your feature request related to a problem? Please describe.**
Build official ARM images for the various gloo components

**Describe the solution you'd like**
ARM compatible images

**Describe alternatives you've considered**
N/A

**Additional context**
[Envoy now provides multi-arch images with arm64 support](https://www.envoyproxy.io/docs/envoy/latest/install/building#arm64-binaries)
[AWS ARM instances](https://aws.amazon.com/ec2/instance-types/a1/)

**Comment by CelsoSantos** [comment]
This should be re-opened as my PR did not solve this issue (it tracks official releases, which still has some work to be done)

**Comment by Sodman** [comment]
Agreed, reopening! We have the makefile changes required to build ARM64 images of Gloo, but we don't have a CI/CD pipeline setup with the ARM worker machines required to build it yet. Our current pipeline is in GKE / Github Actions. We may need to integrate with AWS for this.

Further, for envoy-gloo there is at least one upstream dependency (glibc) which has no official arm64 support.

**Comment by CelsoSantos** [comment]
Indeed. I had to create the glibc image and the frovlad/alpine-glibc images for arm myself: https://github.com/CelsoSantos/alpine-glibc

I recommend Solo forks or recreates this repo in order to handle that

**Comment by murphye** [comment]
Is this on roadmap? Thanks!

**Comment by christian-posta** [comment]
Yah depends on ask, we have the builds for Istio so should be able to get this theoretically if someone needs it.

**Comment by murphye** [comment]
Would be great for Graviton :-)

**Comment by ionutz89** [comment]
Any news about it?

**Comment by WyriHaximus** [comment]
Is there anyway we can help out to move this forward?

**Comment by lgadban** [comment]
@chrisgaun FYI

**Comment by pranaypratyush** [comment]
Need this bad.

**Comment by phenixblue** [comment]
It should be possible to use Docker's `buildx` feature to build multi-arch images without needing access to runners of different architectures.

Example doing this in GitHub Actions: https://github.com/tmobile/magtape/blob/b78170b5b020dd2f0a336d514e48fec5980227cb/.github/workflows/ci-image-build.yaml#L62

**Comment by chrisgaun** [comment]
We have done this with Istio btw and found that there is much high CPU usage for TLS handshakes in ARM vs Intel/AMD. So for those looking to save money using Graviton it would depend on the number of TLS connections.

**Comment by CelsoSantos** [comment]
@chrisgaun would you have some more information to share? Could you eventually do a comparison for the same number of connections how that affect performance? (CPU/RAM/Network, etc). I'd be very interested in learning more about it

**Comment by phenixblue** [comment]
I know a lot of our use case behind this is more for local development so performance differences aren't as much a concern (still a concern if there's a drastic difference). A lot of our development teams use Apple laptops and those will be moving to arm64 as an only option before too long. We like to have a full k8s/gloo deployment in a local environment on the laptops and that's going to be impossible/difficult without upstream support for arm64

**Comment by WyriHaximus** [comment]
To add to the use case list:  I want to use Gloo on my home lab k8s cluster as the ingress service. Doubt I'll see many TLS connections so that's not a major issue for me.

**Comment by chrisgaun** [comment]
Understood understood.

**Comment by ac427** [comment]
@chrisgaun , any new update?

**Comment by scags9876** [comment]
I'm going to second what @phenixblue stated.  Exact same use case for us.   Any updates on this issue being addressed?

**Comment by abebars** [comment]
any updates on this issue?

**Comment by chrisgaun** [comment]
Let me check with engineering. Most run on M1 macs.

**Comment by chrisgaun** [comment]
Everyone. This seems to be a limitation on using Google cloud. They did not have ARM machines until very recently and even now it is very limited supply. 

For Gloo Mesh we use an (expensive) self hosted AWS EKS cluster as a runner to build ARM images. Google says the ARM machines will be GA in "early 2023." Until then we can do one offs for customers potentially but more general pipeline still is a bit out.

**Comment by WyriHaximus** [comment]
You don't need ARM machines to build ARM images tho. All you need is docker with qemu to do a build or buildx to create the images. (It is slow to build...)

**Comment by alex-berger** [comment]
Any progress/plans on this? Currently, `gloo-edge` is the last `amd64` workload in our clusters and (for cost optimization purposes) I wold love to have `arm64` only. 

From what I understand this would imply 
- providing multi architecture OCI manifest (supporting `linux/arm64` and `linux/amd64`) for `quay.io/solo-io/gloo`
- providing multi architecture OCI manifest (supporting `linux/arm64` and `linux/amd64`) for `quay.io/solo-io/gloo-envoy-wrapper`

Maybe I can help with a PR if this is not too complex :-), any pointers and hints are highly appreciated.

**Comment by alex-berger** [comment]
Not stale, still relevant.

**Comment by BeauSelisker28** [comment]
Adding Pegasystems as another one of our high-value customers seeking this. They are hoping we introduce Graviton based images for Gloo Edge so they can better optimize compute and reduce costs.

**Comment by BeauSelisker28** [comment]
Another customer cited seeking Graviton based images for Gloo Edge.

**Comment by nrjpoddar** [comment]
@nfuden to look at this issue for scope/estimation.

Work with @rpunia1 on it.

**Comment by rpunia1** [comment]
PRs 
https://github.com/solo-io/envoy-gloo-ee/pull/830
https://github.com/solo-io/envoy-gloo-ee/pull/849

New envoy-gloo-ee release has been published for version v1.31.2-patch3 after [PR](https://github.com/solo-io/envoy-gloo-ee/pull/849)  got merged
[Release Job](https://github.com/solo-io/envoy-gloo-ee/actions/runs/11710487582/job/32616886410) , commit sha: fbee91b3b2b8006ee2ecee5374e43e7dfee42721
### binaries
[standard, fips](https://console.cloud.google.com/storage/browser/artifacts.solo.io/envoy/v1.31.2-patch3?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&project=gloo-ee&prefix=&forceOnObjectsSortingFiltering=false)
[asan](https://console.cloud.google.com/storage/browser/artifacts.solo.io/envoy?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%257B_22k_22_3A_22_22_2C_22t_22_3A10_2C_22v_22_3A_22_5C_22fbee91b3b2b8006ee2ecee5374e43e7dfee42721%252F_5C_22_22_2C_22s_22_3Atrue%257D%255D%22))&project=gloo-ee&prefix=&forceOnObjectsSortingFiltering=true)
### images
[standard, fips](https://console.cloud.google.com/storage/browser/artifacts.solo.io/envoy/v1.31.2-patch3;tab=objects?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&project=gloo-ee&prefix=&forceOnObjectsSortingFiltering=false)
[asan](https://console.cloud.google.com/storage/browser/artifacts.solo.io/envoy?project=gloo-ee&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%257B_22k_22_3A_22_22_2C_22t_22_3A10_2C_22v_22_3A_22_5C_22fbee91b3b2b8006ee2ecee5374e43e7dfee42721%252F_5C_22_22_2C_22s_22_3Atrue%257D%255D%22))&prefix=&forceOnObjectsSortingFiltering=true)

**Comment by WyriHaximus** [comment]
Are both the PR and GC console internal only?

**Comment by prathyk** [comment]
Any update on this issue? Is there a place where I can pick gloo arm images?

**Comment by nfuden** [comment]
Hey sorry that this has been going on for so long. We are looking to get onto arm soon but this repo is also determining how it wants to pull its envoy images / expand on envoy functionality. 
Now that there are public runners for github actions we hope that this is something we can adopt. 
historically emulation for the builds of what was the envoy extensions were somewhat painful which made contributions less likely

**Comment by mgnisia** [comment]
Good morning, can this issue somehow support e.g. providing ARM infrastructure for the CICD pipelines?

**Comment by npolshakova** [comment]
kgateway + agentgateway now supports arm64! 🎉 https://github.com/kgateway-dev/kgateway/pkgs/container/agentgateway-controller/626352977?tag=v2.2.0-main

**Comment by mikkoc** [comment]
> kgateway + agentgateway now supports arm64! 🎉 https://github.com/kgateway-dev/kgateway/pkgs/container/agentgateway-controller/626352977?tag=v2.2.0-main

Thanks Nina. Do you have plans to publish an arm64 version of the `envoy-wrapper` image?
Trying to bring up this Service still fails because it pulls the envoy-wrapper image, which does not run on our ARM nodes.
```
apiVersion: gateway.kgateway.dev/v1alpha1
kind: GatewayParameters
metadata:
  name: kgw-params
  namespace: kgateway-system
spec:
  kube:
    service:
      type: NodePort
      ports:
        - port: 80
          nodePort: 31339
```

**Comment by manjitkumar** [comment]
I've tried to deploy kgateway onto an EKS cluster (1.34) with ARM64 only nodes and found that all kgateway `envoy-wrapper` ARM64 images are failing to start and have AMD64 base system binaries (/bin/sh, /bin/dash), even though they're tagged as ARM64 and contain ARM64 Envoy binaries. This causes the "exec format error" on AWS Graviton (ARM64) nodes.

  Tested Images (All Failed):

  - cr.kgateway.dev/kgateway-dev/envoy-wrapper:v2.2.0-main@sha256:2cde867cb51d5d79c6c4e32bcf06ddabf9a3f8bbda78e605e516f99dcb363a93
  - ghcr.io/kgateway-dev/envoy-wrapper:v2.1.2-arm64
  - ghcr.io/kgateway-dev/envoy-wrapper:v2.1.2

**Comment by lgadban** [comment]
Apparently we are hitting the limit of linked PRs :)

Non-exhaustive list of related PRs:
- https://github.com/kgateway-dev/kgateway/pull/12950
- https://github.com/kgateway-dev/kgateway/pull/13307 
- https://github.com/kgateway-dev/kgateway/pull/13319
- https://github.com/kgateway-dev/kgateway/pull/13406 
- https://github.com/kgateway-dev/kgateway/pull/13356
- https://github.com/kgateway-dev/kgateway/pull/13307
- https://github.com/kgateway-dev/kgateway/pull/13411
- https://github.com/kgateway-dev/kgateway/pull/13386
- https://github.com/kgateway-dev/kgateway/pull/13410
- https://github.com/kgateway-dev/kgateway/pull/13426
- https://github.com/kgateway-dev/kgateway/pull/13448
- https://github.com/kgateway-dev/kgateway/pull/13437

**Comment by lgadban** [comment]
Thank you everyone for the patience on this issue, it is with great honor that I can announce this issue is finally completed!  🎉🎉 

As of the [v2.2.0 release of kgateway](https://github.com/kgateway-dev/kgateway/releases/tag/v2.2.0) we have proper arm64 support for the envoy dataplane.

Now, we can finally close this issue that I opened back when we still in the early days of our 1.x project, Gloo Edge.
It only took us five and a half years 😉 

Please give it a try and let us know how it goes!

**Comment by nfuden** [comment]
Amazing stuff congratulations everyone!

**Comment by lgadban** [comment]
thanks for your contributions on getting rustformations off the ground as well, @nfuden !
